File: Examples/test-suite/perl5/char_strings_runme.pl
use strict;
use warnings;
use Test::More tests => 29;
BEGIN { use_ok('char_strings') }
require_ok('char_strings');

my $CPLUSPLUS_MSG = "A message from the deep dark world of C++, where anything is possible.";
my $OTHERLAND_MSG = "Little message from the safe world.";

# get functions
is(char_strings::GetCharHeapString(), $CPLUSPLUS_MSG, 'char get 1');
char_strings::DeleteCharHeapString();

is(char_strings::GetConstCharProgramCodeString(), $CPLUSPLUS_MSG, 'char get 2');

is(char_strings::GetCharStaticString(), $CPLUSPLUS_MSG, 'char get 3');

is(char_strings::GetCharStaticStringFixed(), $CPLUSPLUS_MSG, 'char get 4');

is(char_strings::GetConstCharStaticStringFixed(), $CPLUSPLUS_MSG, 'char get 5');

# set functions - Set(OTHERLAND_MSG . number, number) must return true
ok(char_strings::SetCharHeapString($OTHERLAND_MSG . 1, 1), 'char set 1');
ok(char_strings::SetCharStaticString($OTHERLAND_MSG . 2, 2), 'char set 2');
ok(char_strings::SetCharArrayStaticString($OTHERLAND_MSG . 3, 3), 'char set 3');
ok(char_strings::SetConstCharHeapString($OTHERLAND_MSG . 4, 4), 'char set 4');
ok(char_strings::SetConstCharStaticString($OTHERLAND_MSG . 5, 5), 'char set 5');
ok(char_strings::SetConstCharArrayStaticString($OTHERLAND_MSG . 6, 6), 'char set 6');
ok(char_strings::SetCharConstStaticString($OTHERLAND_MSG . 7, 7), 'char set 7');
ok(char_strings::SetConstCharConstStaticString($OTHERLAND_MSG . 8, 8), 'char set 8');
ok(char_strings::SetConstCharTypedefString($OTHERLAND_MSG . 9, 9), 'char set 9');

# get set functions (ping pong round trip)
{
  my $ping = $OTHERLAND_MSG . 11;
  is(char_strings::CharPingPong($ping), $ping, 'PingPong 1');
}
{
  my $ping = $OTHERLAND_MSG . 12;
  is(char_strings::CharArrayPingPong($ping), $ping, 'PingPong 2');
}
# CharArrayDimsPingPong takes char[16], so the string must fit in 15 chars + nul
{
  my $ping = "PingPong 13";
  is(char_strings::CharArrayDimsPingPong($ping), $ping, 'PingPong 3');
}

# variables
{
  my $expected = $OTHERLAND_MSG . 21;
  $char_strings::global_char = $expected;
  is($char_strings::global_char, $expected, 'variables 1 (global_char set+read)');
}
# global_char_array1 is read-only in Perl (char[] has no varin typemap), so only read it
is($char_strings::global_char_array1, $CPLUSPLUS_MSG, 'variables 2 (global_char_array1 read)');
{
  my $expected = $OTHERLAND_MSG . 23;
  $char_strings::global_char_array2 = $expected;
  is($char_strings::global_char_array2, $expected, 'variables 3 (global_char_array2 set+read)');
}
is($char_strings::global_const_char, $CPLUSPLUS_MSG, 'variables 4 (global_const_char read-only)');
is($char_strings::global_const_char_array1, $CPLUSPLUS_MSG, 'variables 5 (global_const_char_array1 read-only)');
is($char_strings::global_const_char_array2, $CPLUSPLUS_MSG, 'variables 6 (global_const_char_array2 read-only)');

# char *& tests
is(char_strings::GetCharPointerRef(), $CPLUSPLUS_MSG, 'char *& get 1');
ok(char_strings::SetCharPointerRef($OTHERLAND_MSG . 31, 31), 'char *& set 2');
is(char_strings::GetConstCharPointerRef(), $CPLUSPLUS_MSG, 'char *& get 3');
ok(char_strings::SetConstCharPointerRef($OTHERLAND_MSG . 33, 33), 'char *& set 4');

File: Examples/test-suite/octave/char_strings_runme.m
# do not dump Octave core
if exist("crash_dumps_octave_core", "builtin")
  crash_dumps_octave_core(0);
endif

char_strings

CPLUSPLUS_MSG = "A message from the deep dark world of C++, where anything is possible.";
OTHERLAND_MSG = "Little message from the safe world.";

count = 5;

# get functions
for i = 0:count-1
  if (!strcmp(char_strings.GetCharHeapString(), CPLUSPLUS_MSG))
    error("Test char get 1 failed");
  endif
  char_strings.DeleteCharHeapString();
endfor

for i = 0:count-1
  if (!strcmp(char_strings.GetConstCharProgramCodeString(), CPLUSPLUS_MSG))
    error("Test char get 2 failed");
  endif
  char_strings.DeleteCharHeapString();
endfor

for i = 0:count-1
  if (!strcmp(char_strings.GetCharStaticString(), CPLUSPLUS_MSG))
    error("Test char get 3 failed");
  endif
endfor

for i = 0:count-1
  if (!strcmp(char_strings.GetCharStaticStringFixed(), CPLUSPLUS_MSG))
    error("Test char get 4 failed");
  endif
endfor

for i = 0:count-1
  if (!strcmp(char_strings.GetConstCharStaticStringFixed(), CPLUSPLUS_MSG))
    error("Test char get 5 failed");
  endif
endfor

# set functions
for i = 0:count-1
  if (!char_strings.SetCharHeapString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 1 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetCharStaticString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 2 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetCharArrayStaticString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 3 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetConstCharHeapString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 4 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetConstCharStaticString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 5 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetConstCharArrayStaticString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 6 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetCharConstStaticString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 7 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetConstCharConstStaticString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 8 failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetConstCharTypedefString([OTHERLAND_MSG num2str(i)], i))
    error("Test char set 9 failed");
  endif
endfor

# get set function (ping pong)
for i = 0:count-1
  ping = [OTHERLAND_MSG num2str(i)];
  pong = char_strings.CharPingPong(ping);
  if (!strcmp(ping, pong))
    error("Test PingPong 1 failed");
  endif
endfor

for i = 0:count-1
  ping = [OTHERLAND_MSG num2str(i)];
  pong = char_strings.CharArrayPingPong(ping);
  if (!strcmp(ping, pong))
    error("Test PingPong 2 failed");
  endif
endfor

# CharArrayDimsPingPong takes a fixed char[16] buffer, so the string must fit in 16 bytes
for i = 0:count-1
  ping = ["PingPong" num2str(i)];
  pong = char_strings.CharArrayDimsPingPong(ping);
  if (!strcmp(ping, pong))
    error("Test PingPong 3 failed");
  endif
endfor

# variables
for i = 0:count-1
  expected = [OTHERLAND_MSG num2str(i)];
  char_strings.cvar.global_char = expected;
  if (!strcmp(char_strings.cvar.global_char, expected))
    error("Test variables 1 failed");
  endif
endfor

# global_char_array1 is char[] with no dimension => read-only in Octave (VARIN typemap
# undefined, see %warnfilter in char_strings.i). It retains its initial value.
for i = 0:count-1
  if (!strcmp(char_strings.cvar.global_char_array1, CPLUSPLUS_MSG))
    error("Test variables 2 failed");
  endif
endfor

for i = 0:count-1
  expected = [OTHERLAND_MSG num2str(i)];
  char_strings.cvar.global_char_array2 = expected;
  if (!strcmp(char_strings.cvar.global_char_array2, expected))
    error("Test variables 3 failed");
  endif
endfor

for i = 0:count-1
  if (!strcmp(char_strings.cvar.global_const_char, CPLUSPLUS_MSG))
    error("Test variables 4 failed");
  endif
endfor

for i = 0:count-1
  if (!strcmp(char_strings.cvar.global_const_char_array1, CPLUSPLUS_MSG))
    error("Test variables 5 failed");
  endif
endfor

for i = 0:count-1
  if (!strcmp(char_strings.cvar.global_const_char_array2, CPLUSPLUS_MSG))
    error("Test variables 6 failed");
  endif
endfor

# char *& tests
for i = 0:count-1
  if (!strcmp(char_strings.GetCharPointerRef(), CPLUSPLUS_MSG))
    error("Test char pointer ref get failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetCharPointerRef([OTHERLAND_MSG num2str(i)], i))
    error("Test char pointer ref set failed");
  endif
endfor

for i = 0:count-1
  if (!strcmp(char_strings.GetConstCharPointerRef(), CPLUSPLUS_MSG))
    error("Test const char pointer ref get failed");
  endif
endfor

for i = 0:count-1
  if (!char_strings.SetConstCharPointerRef([OTHERLAND_MSG num2str(i)], i))
    error("Test const char pointer ref set failed");
  endif
endfor
